src/k8s/pkg/snap/snap.go

```diff
@@ -0,0 +1,102 @@
+package snap
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/canonical/k8s/pkg/utils"
+)
+
+// snap implements the Snap interface.
+type snap struct {
+	snapDir       string
+	snapDataDir   string
+	snapCommonDir string
+}
+
+// NewSnap creates a new interface with the K8s snap.
+// NewSnap accepts the $SNAP, $SNAP_DATA and $SNAP_COMMON, directories
+func NewSnap(snapDir, snapDataDir, snapCommonDir string, options ...func(s *snap)) Snap {
+	s := &snap{
+		snapDir:       snapDir,
+		snapDataDir:   snapDataDir,
+		snapCommonDir: snapCommonDir,
+	}
+
+	for _, opt := range options {
+		opt(s)
+	}
+	return s
+}
+
+var snapInstance Snap
+var lock = &sync.Mutex{}
```

Comment:
I see why we do this here, but I don't believe there is any reason to go with a singleton object. I think at this point it's OK to initialize new instances as needed, there is not much initialization happening, just a struct being created.

The right way to approach is that we create an instance when we start then ensure to pass it around as needed, not a singleton object.

I would recommend just removing this, we can keep `snap.Instance()` as a temporary easy solution.

Reply:
fair - i'll change it